utils/merge.go

```diff
@@ -0,0 +1,147 @@
+//   - if maps are different, they will be merged according to the following rules:
+//       - keys in base are added first
+//       - keys in patch are added after, overwriting existing keys with values from patch
+//       - reference values (eg. slice/ptr/map/chan) will not be cloned
+//   - channel values are not supported at the moment
+//
+// Usage: callers need to cast the returned interface back into the original type, eg:
+// func mergeTestStruct(base, patch *testStruct) (*testStruct, error) {
+//     ret, err := merge(base, patch)
+//     if err != nil {
+//         return nil, err
+//     }
+//     retTS := ret.(testStruct)
+//     return &retTS, nil
+// }
+func Merge(base interface{}, patch interface{}) (interface{}, error) {
+	if reflect.TypeOf(base) != reflect.TypeOf(patch) {
+		return nil, fmt.Errorf("cannot merge different types. base type: %s, patch type: %s",
+			reflect.TypeOf(base), reflect.TypeOf(patch))
+	}
+
+	baseType := reflect.TypeOf(base)
+	baseVal := reflect.ValueOf(base)
+	patchVal := reflect.ValueOf(patch)
+	if baseType.Kind() == reflect.Ptr {
+		baseType = baseType.Elem()
+		baseVal = baseVal.Elem()
+		patchVal = patchVal.Elem()
+	}
+
+	ret := reflect.New(baseType)
+
+	val, ok := merge(baseVal, patchVal)
+	if ok {
+		ret.Elem().Set(val)
+	}
+	return ret.Elem().Interface(), nil
+}
+
+// merge recursively merges patch into base and returns the new struct, ptr, slice/map, or value
+func merge(base, patch reflect.Value) (reflect.Value, bool) {
+	baseType := base.Type()
+
+	switch baseType.Kind() {
+	case reflect.Struct:
+		merged := reflect.New(baseType).Elem()
+		for i := 0; i < base.NumField(); i++ {
+			if !merged.Field(i).CanSet() {
+				continue
+			}
+			val, ok := merge(base.Field(i), patch.Field(i))
+			if ok {
+				merged.Field(i).Set(val)
+			}
+		}
+		return merged, true
+
+	case reflect.Ptr:
+		mergedPtr := reflect.New(baseType.Elem())
+		if base.IsNil() && patch.IsNil() {
+			return mergedPtr, false
+		} else if base.IsNil() {
+			mergedPtr.Elem().Set(patch.Elem())
+			return mergedPtr, true
+		} else if patch.IsNil() {
+			mergedPtr.Elem().Set(base.Elem())
+			return mergedPtr, true
+		}
+		val, ok := merge(base.Elem(), patch.Elem())
+		if ok {
+			mergedPtr.Elem().Set(val)
+		}
+		return mergedPtr, true
+
+	case reflect.Slice:
+		// merge with these rules:
+		// - start with elements from base (if any)
+		// - then for each item, patch will overwrite if different
+		// - and then add the extra items from patch (if any)
+		if base.IsNil() && patch.IsNil() {
+			return reflect.Zero(baseType), false
+		}
+		var maxLen int
+		if base.Len() > patch.Len() {
+			maxLen = base.Len()
+		} else {
+			maxLen = patch.Len()
+		}
+		merged := reflect.MakeSlice(baseType, maxLen, maxLen)
+		if !base.IsNil() {
+			reflect.Copy(merged, base)
+		}
+		if !patch.IsNil() {
+			reflect.Copy(merged, patch)
+		}
+		return merged, true
+
+	case reflect.Map:
+		// for now, maps are merged in a very rudimentary way:
```

Comment:
In general, I can imagine these merge semantics being more useful than the slice semantics above, but a few thoughts if we keep this:
* Should `nil` values in `patch` be ignored instead of clobbering `non-nil` values in `base`?
* If both `base` and `patch` are non-nil, it seems like we could safely clone reference values by recursing again:
```
value, ok := merge(base.MapIndex(key), patch.MapIndex(key))
if ok {
    merged.SetMapIndex(key, value)
}
```
* If one of `base` or `patch` is `nil`, I think we can still safely clone:
```
value, ok := merge(reflect.Zero(baseType), baseOrPatchWhicheverIsNotNil.MapIndex(key))
if ok {
    merged.SetMapIndex(key, value)
}
```

Reply:
Good catch! I've changed it now to reflect this better behavior.